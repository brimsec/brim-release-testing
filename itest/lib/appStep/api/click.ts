import {Application} from "spectron"

import {retryUntil} from "../../control"
import {LOG} from "../../log"

import logStep from "../util/logStep"

const waitForClickable = async (app: Application, selector: string) => {
  // In testing, it's been shown than there is no need to scroll to
  // elements to make them be visible, as long as the element's
  // container allows scrolling. However, some Internet searches suggest
  // scrolling to the element before trying to click in order to avoid
  // problems like those described in
  // https://github.com/brimsec/brim/issues/668
  await logStep(`wait for element to exist: "${selector}"`, () =>
    app.client.waitForExist(selector)
  )
  await logStep(`wait for element to be visible: "${selector}"`, () =>
    app.client.waitForVisible(selector)
  )
  return logStep(`scroll to: "${selector}"`, () => app.client.scroll(selector))
}

export const click = (app: Application, selector: string) =>
  logStep(`click on selector "${selector}"`, async () => {
    await waitForClickable(app, selector)
    try {
      return retryUntil(
        () => app.client.click(selector),
        (success) => success
      )
    } catch (e) {
      LOG.debug("trying to execute script for click: " + e)
      return app.client.selectorExecute(selector, (elem) => {
        elem.click()
      })
    }
  })

export const rightClick = (app: Application, selector: string) =>
  logStep(`right-click on selector "${selector}"`, async () => {
    await waitForClickable(app, selector)
    try {
      return retryUntil(
        () => app.client.rightClick(selector),
        (success) => success
      )
    } catch (e) {
      LOG.debug("trying to execute script for rightClick: " + e)
      return app.client.selectorExecute(selector, (elem) => {
        elem.rightClick()
      })
    }
  })

export const waitForClickableButtonAndClick = async (
  app: Application,
  selector: string
) => {
  await waitForClickable(app, selector)
  // In addition to waitForClickable above, buttons must also be enabled.
  await logStep(`wait for button ${selector} to be enabled`, () =>
    retryUntil(
      () => app.client.getAttribute(selector, "disabled"),
      (isDisabled) => !isDisabled
    )
  )

  // We can use app.client.click() here because we've done the necessary
  // waiting.
  try {
    return retryUntil(
      () => app.client.click(selector),
      (success) => success
    )
  } catch (e) {
    LOG.debug("trying to execute script for click: " + e)
    return app.client.selectorExecute(selector, (elem) => {
      elem.click()
    })
  }
}
