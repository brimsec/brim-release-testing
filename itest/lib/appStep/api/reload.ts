import {Application} from "spectron"

import logStep from "../util/logStep"
import {retry} from "../../control"

export default (app: Application) =>
  // Issue a browser window reload(). This is flaky in the Spectron,
  // Chromedriver stack, but appears be contained within the method.
  // It seems safe to retry.
  // https://github.com/brimsec/brim/issues/878
  // https://github.com/electron-userland/spectron/issues/493
  logStep("app reload", () => retry(() => app.browserWindow.reload(), 60, 1000))
