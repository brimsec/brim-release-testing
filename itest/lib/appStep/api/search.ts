import {Application} from "spectron"

import logStep from "../util/logStep"
import waitForResults from "../util//waitForResults"
import {selectors} from "../../../../src/js/test/integration"

export const runSearch = async (app: Application, searchText: string) => {
  // Run a search and wait for results to appear. Do not return results.
  // This is suitable for setting up the app to do something else
  // after receiving results, like a context menu test.
  await logStep("write to main search", () =>
    app.client
      .waitForVisible(selectors.search.input)
      .then(() => app.client.setValue(selectors.search.input, searchText))
  )
  await logStep("click the search button", () => app.client.keys("Enter"))
  return waitForResults(app)
}

export const getResults = async (
  app: Application,
  includeHeaders = true
): Promise<string[]> => {
  // Return search current search results. This is useful for getting
  // the current results by doing a search through other means, like a
  // context menu-built search.

  const searchDisplayHeaders = async (app: Application): Promise<string[]> => {
    // This stinks. We have to use getHTML because headers that are off the
    // screen return as empty strings if you use getText. This isn't required of
    // actual results.
    // See http://v4.webdriver.io/api/property/getText.html
    // app.browserWindow.maximize() fixes the problem on my laptop but not CI.
    // But what we get back includes the width which can be non-deterministic:
    // <div class="header-cell" style="width: 192px;">ts<div class="col-resizer"></div></div>
    // That style width will vary on my laptop vs. CI.
    // The hack is to split this and extract just the text.
    const _trim = (s: string) => s.split(">")[1].split("<")[0]

    let headers = await logStep("get search headers", () =>
      app.client
        .waitForVisible(selectors.viewer.headers)
        .then(() => app.client.getHTML(selectors.viewer.headers))
    )
    if (typeof headers === "string") {
      headers = [headers]
    }
    return headers.map((h) => _trim(h))
  }

  const searchResults = await logStep("get search records", async () => {
    await app.client.waitForVisible(selectors.viewer.results)
    return app.client.getText(selectors.viewer.results)
  })

  let headers
  if (
    includeHeaders &&
    (await app.client.isVisible(selectors.viewer.headers))
  ) {
    headers = await searchDisplayHeaders(app)
  } else {
    headers = []
  }
  return headers.concat(searchResults)
}

export default async (app: Application, searchText: string) => {
  // Run a search and return results.
  await runSearch(app, searchText)
  return getResults(app)
}
