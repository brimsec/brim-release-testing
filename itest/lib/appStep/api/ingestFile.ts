import {Application} from "spectron"
import path from "path"

import {testDataDir} from "../../env"
import {retryUntil} from "../../control"
import {click} from "./click"
import logStep from "../util/logStep"
import {LOG} from "../../log"
import {selectors} from "../../../../src/js/test/integration"

export default async (app: Application, file: string) => {
  // Ingest a file inside the itest/testdata directory
  await click(app, ".add-tab")

  await logStep("wait for pcap file input", () =>
    app.client.waitForVisible(selectors.ingest.filesButton)
  )
  await logStep("choose file", () =>
    app.client.chooseFile(
      selectors.ingest.filesInput,
      path.normalize(path.join(testDataDir(), file))
    )
  )

  try {
    await logStep("wait for ingest to start", () =>
      retryUntil(
        () => app.client.isExisting(selectors.status.ingestProgress),
        (ingesting) => ingesting === true,
        100,
        100
      )
    )
  } catch {
    LOG.debug(
      "ingest never appeared; let's hope it finished and ended before we could observe it"
    )
  }

  return logStep("wait for ingest to finish", () =>
    retryUntil(
      () => app.client.isExisting(selectors.status.ingestProgress),
      (ingesting) => ingesting === false
    )
  )
}
