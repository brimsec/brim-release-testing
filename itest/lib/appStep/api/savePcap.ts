import {Application} from "spectron"

import {retryUntil} from "../../control"
import {waitForClickableButtonAndClick} from "./click"
import logStep from "../util/logStep"
import {selectors} from "../../../../src/js/test/integration"

export default async (app: Application) => {
  await waitForClickableButtonAndClick(app, selectors.pcaps.button)
  return logStep("wait for a download to finish", async () => {
    await app.client.waitForVisible(selectors.downloadMessage)
    return retryUntil(
      () => app.client.getText(selectors.downloadMessage),
      (text) => text == "Download Complete" || text.includes("Download error")
    )
  })
}

export const pcapsDir = (app: Application) =>
  app.electron.remote.app.getPath("temp")
