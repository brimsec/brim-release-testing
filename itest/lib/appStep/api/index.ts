import {click, rightClick} from "./click"
import ingestFile from "./ingestFile"
import reload from "./reload"
import savePcap from "./savePcap"
import search from "./search"
import setSpan from "./setSpan"
import startApp from "./startApp"
import takeScreenshot from "./takeScreenshot"

export default {
  click,
  rightClick,
  ingestFile,
  reload,
  savePcap,
  search,
  setSpan,
  startApp,
  takeScreenshot
}
