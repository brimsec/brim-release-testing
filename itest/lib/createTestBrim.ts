import {Locator} from "../../src/js/test/createLocator"
import appStep from "./appStep/api"
// TODO in a future PR: remove direct logStep uses here.
import logStep from "./appStep/util/logStep"
import newAppInstance from "./newAppInstance"
import {retryUntil} from "./control"
import lib from "../../src/js/lib"

export default (name: string) => {
  let app
  let testIdx = 0

  beforeAll(async () => {
    app = await newAppInstance(name, ++testIdx)
    await appStep.startApp(app)
  })

  afterAll(async () => {
    if (app && app.isRunning()) await app.stop()
  })

  return {
    getApp() {
      return app
    },

    ingest(file: string) {
      return appStep.ingestFile(app, file)
    },

    mockSaveDialog(value: {
      canceled: boolean
      filePath: string | null | undefined
    }) {
      return app.mainProcess.emit(
        "spectron:mock",
        "windows:showSaveDialog",
        value
      )
    },

    setValue(locator: Locator, value: string) {
      return logStep(`set input ${locator.css} to ${value}`, () =>
        app.client
          .waitForVisible(locator.css)
          .then(() => app.client.setValue(locator.css, value))
      )
    },

    getText(locator: Locator) {
      return logStep(`get text from ${locator.css}`, async () => {
        await app.client.waitForVisible(locator.css)
        return await app.client.getText(locator.css)
      })
    },

    clickAppMenuItem(id: string) {
      return app.mainProcess.emit("spectron:clickAppMenuItem", id)
    },

    search(input: string) {
      return appStep.search(app, input)
    },

    click(locator: Locator) {
      return appStep.click(app, locator.css)
    },

    waitForText(locator: string, regex: RegExp) {
      return retryUntil(
        () => app.client.getText(locator),
        (s) => regex.test(s)
      )
    },

    waitUntil(fn: () => boolean, opts?: Object) {
      return app.client.waitUntil(fn, opts)
    },

    isVisible(locator: Locator) {
      return app.client.isVisible(locator.css)
    },

    isNotVisible(locator: Locator) {
      return app.client.isVisible(locator.css).then((v) => !v)
    },

    wait(ms: number) {
      return lib.sleep(ms)
    }
  }
}
