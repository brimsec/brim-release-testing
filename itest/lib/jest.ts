import {Application} from "spectron"

import appStep from "./appStep/api"
import {selectors} from "../../src/js/test/integration"
import {LOG} from "./log"

export const TestTimeout = 300000
// https://jestjs.io/docs/en/troubleshooting#unresolved-promises
// https://jestjs.io/docs/en/jest-object#jestsettimeouttimeout
jest.setTimeout(TestTimeout)

export const handleError = async (
  app: Application,
  initialError: Error,
  done: any
) => {
  appStep.takeScreenshot(app)
  let realError = undefined
  let notificationError = undefined
  LOG.error(`handleError: Test hit exception: ${initialError.message}`)
  LOG.info("handleError: Looking for any desktop app notifications")
  try {
    notificationError = await app.client.getHTML(selectors.notification, false)
  } catch (e) {
    notificationError = undefined
  }
  if (notificationError) {
    realError = new Error(
      "App notification '" +
        notificationError +
        "' (initial error: '" +
        initialError +
        "'"
    )
  } else {
    LOG.info("handleError: desktop app notification not found")
    realError = initialError
  }
  done.fail && done.fail(realError)
}

// Put tests in a wrapper that prints what test is being run. This is
// to eagerly print to the console where the run is so that hangs and other
// failures are easier to diagnose. Without this, Jest is pretty silent and
// it's hard to know what went wrong in a CI run. This MUST be paired with jest
// --verbose to work properly, otherwise LOG.info() calls seem to get
// buffered and printed later.
export const stdTest = (
  descr: string,
  f: (done: any) => void | Promise<void>,
  timeout: number = TestTimeout
) => {
  test(
    descr,
    (done) => {
      LOG.info(`Running test: ${descr}`)
      f(done)
    },
    timeout
  )
}
