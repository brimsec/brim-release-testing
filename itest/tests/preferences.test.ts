import path from "path"

import {
  defaultModalButton,
  jsonTypeConfigInput
} from "../../src/js/test/locators"
import {getResults} from "../lib/appStep/api/search"
import {stdTest} from "../lib/jest"
import {testDataDir} from "../lib/env"
import createTestBrim from "../lib/createTestBrim"

const config = path.join(testDataDir(), "custom-schema.json")

describe("Preferences Modal Tests", () => {
  const brim = createTestBrim("preferences.test")

  stdTest("setting a custom json type config", async (done) => {
    brim.clickAppMenuItem("preferences")
    await brim.setValue(jsonTypeConfigInput, config)
    await brim.click(defaultModalButton)
    await brim.ingest("custom-sample.ndjson")
    await brim.search(
      "_path=conn proto=tcp | cut ts, src_ip, src_port, dst_ip, dst_port, proto | sort ts | head 10"
    )

    expect(await getResults(brim.getApp())).toMatchSnapshot()
    done()
  })
})
