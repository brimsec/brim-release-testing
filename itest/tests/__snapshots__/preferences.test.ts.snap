// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Preferences Modal Tests setting a custom json type config 1`] = `
Array [
  "ts",
  "src_ip",
  "src_port",
  "dst_ip",
  "dst_port",
  "proto",
  "2020-02-25T16:03:05Z",
  "192.168.1.110",
  "56625",
  "172.217.9.142",
  "443",
  "tcp",
  "2020-02-25T16:03:06Z",
  "192.168.1.110",
  "57591",
  "172.217.9.142",
  "443",
  "tcp",
  "2020-02-25T16:03:07Z",
  "192.168.1.110",
  "55635",
  "18.246.31.137",
  "443",
  "tcp",
  "2020-02-25T16:03:07Z",
  "192.168.1.110",
  "55747",
  "13.52.5.22",
  "443",
  "tcp",
  "2020-02-25T16:03:08Z",
  "192.168.1.110",
  "55344",
  "52.37.243.173",
  "443",
  "tcp",
  "2020-02-25T16:03:08Z",
  "192.168.1.110",
  "55346",
  "52.37.243.173",
  "443",
  "tcp",
  "2020-02-25T16:03:08Z",
  "192.168.1.110",
  "57326",
  "173.194.201.189",
  "443",
  "tcp",
  "2020-02-25T16:03:08Z",
  "192.168.1.110",
  "57487",
  "192.30.253.125",
  "443",
  "tcp",
  "2020-02-25T16:03:08Z",
  "192.168.1.110",
  "55354",
  "52.37.243.173",
  "443",
  "tcp",
  "2020-02-25T16:03:10Z",
  "192.168.1.110",
  "57332",
  "64.233.179.189",
  "443",
  "tcp",
]
`;
